### DISCUSSION TITLE
Connection Handling in Express

### DISCUSSION THREAD

**Discussion by bitsmyth** [prisma/prisma#4825] [Q&A] [answered]
Let me tell you that I am using **prisma with nextjs**, it works pretty solid - _now_.

 I want to tell you all to **forgett about** the **serverless** environment that [nextjs is providing](https://nextjs.org/docs/api-routes/introduction), because **each function will open a new connection** to your database. Serverless is good for things you "fire up and forgett", dont do the same mistake like me and use serverless for connecting to your database, even when prisma is telling you it is possible. 

Lessons learned, so I **migrated** my functions to an **express server** and instantly saw an **decrease in open connections**. But from time to time as I monitored the connections I also came across increasing connections even when I [set the limit](https://www.prisma.io/docs/concepts/components/prisma-client/connection-management#connection-pool) at three. That was because some connections kept open while developing and making mistakes - _errors_. 

Prisma have a pretty fair documentation about [connection-management](https://www.prisma.io/docs/concepts/components/prisma-client/connection-management) but I have to figure out how to implement this right into express. And I found that this is the way to go.

```ts
/**
 *    GLOBAL ERROR HANDLING
 *    ! DISCONNECT PRISMA
 */
app.use(async (error, req, res, next) => {
  console.log(error);
  await prisma.$disconnect(); // <- here

  const status = error.statusCode || 500;
  const message = error.message;
  const data = error.data;
  res.status(status).json({ message: message, data: data });
});

app.listen(process.env.SERVER_PORT, function () {
  console.log('Listening on port ' + process.env.SERVER_PORT);
});

/**
 *    Safety measure
 *    ! DISCONNECT PRISMA WHEN WE WROTE BAD CODE
 */
process.on('uncaughtException', async (err) => {
  console.log(err);
  await prisma.$disconnect(); // <- here
});
```

**Comment by nikolasburk** [comment, marked as answer]
Hey @NikoMontana, thanks a lot for raising this!

Generally the setup looks solid, but I have two comments:

- Is there a special reason that you want to disconnect in the global error handling middleware? Usually in that case the server should keep running, so there's not really a reason to disconnect – curious to hear more about your use case here! 
- It also seems like the current `uncaughtException` is dangerous because it overrides the `exit` meaning the app will not exit. See the docs herer: https://nodejs.org/docs/latest-v14.x/api/process.html#process_event_uncaughtexception. You can use this code instead: 
 	```ts
	process.on('uncaughtException', async (err) => {
	  console.error(`Uncaught Exception: ${err.message}`)
	  await prisma.$disconnect();
	  process.exit(1)
	})
	```
  Also note that this shouldn't actually be needed because we already listen to `exit`, `SIGINT`, `SIGUSR1`, `SIGUSR2`, `SIGTERM` events: https://github.com/prisma/prisma/blob/89f1abd71fdbda8d1924b45edd360b396ba57188/src/packages/engine-core/src/NodeEngine.ts#L1248-L1279

**Reply by andremarcondesteixeira** [reply]
If Prisma already listens for application exit, then why does the QuickStart docs recommend taking care of disconnection manually? Do we still need to take care of disconnections?
https://www.prisma.io/docs/getting-started/quickstart

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
